File: selenide-pages-sample/src/test/java/org/brewcode/qa/pages/sample/java/page/DockerGettingStartedMainPage.java
package org.brewcode.qa.pages.sample.java.page;

import com.codeborne.selenide.ElementsCollection;
import com.codeborne.selenide.SelenideElement;
import org.brewcode.qa.pages.annotation.Element;
import org.brewcode.qa.pages.annotation.Page;
import org.brewcode.qa.pages.element.Block;
import org.brewcode.qa.pages.element.Blocks;
import org.brewcode.qa.pages.page.BasePage;

/**
 * Define selenide page via @Page annotation. Must extend BasePage class.
 * 
 * value = "Getting Started" - page name for log message
 * path = "/tutorial/" - expected path. Will check on page open verify, and will append when open this page
 * expectedTitle = "Getting Started" - The title text will check on page open verify
 */
@Page(value = "Getting Started", path = "/tutorial/", expectedTitle = "Getting Started")
public class DockerGettingStartedMainPage extends BasePage<DockerGettingStartedMainPage> {

    /**
     * Create element collection.
     * With alias "Paragraphs" to use in log messages.
     * Search by tag name.
     * 
     * "required = true" - This element must be visible on the page after loading. When you open the page, this element must be visible, otherwise an error will occur
     */
    @Element(value = "Paragraphs", tagName = "h2", required = true)
    public ElementsCollection paragraphs;

    public GettingStartedNavigation gettingStartedNavigation;

    /**
     * The page block defined.
     * Must extend Block class.
     * 
     * Be aware - @Element annotation uses on class instead of field. 
     * But you may use field annotation and field annotation has a higher priority.
     */
    @Element(value = "Getting started navigation", css = "nav.md-nav--primary>ul")
    public static class GettingStartedNavigation extends Block {

        /**
         * Collection of page blocks.
         */
        @Element(value = "Пункты навигации", xpath = "./li")
        public Blocks<NavigationItem> navigationItemList;
    }

    public static class NavigationItem extends Block {

        @Element(value = "Navigation item", tagName = "a", required = true)
        public SelenideElement gettingStartedNavigation;
    }
}

File: selenide-pages-sample/src/test/kotlin/org/brewcode/qa/pages/sample/kotlin/page/DockerGettingStartedMainPage.kt
package org.brewcode.qa.pages.sample.kotlin.page

import com.codeborne.selenide.Condition
import com.codeborne.selenide.ElementsCollection
import com.codeborne.selenide.SelenideElement
import io.qameta.allure.Step
import org.brewcode.qa.pages.annotation.Element
import org.brewcode.qa.pages.annotation.Page
import org.brewcode.qa.pages.element.Block
import org.brewcode.qa.pages.element.Blocks
import org.brewcode.qa.pages.page.BasePage

@Page(value = "Getting Started", path = "/tutorial/", expectedTitle = "Getting Started")
class DockerGettingStartedMainPage : BasePage<DockerGettingStartedMainPage>() {

    @Element(value = "Paragraphs", tagName = "h2")
    var paragraphs: ElementsCollection? = null

    lateinit var dockerRunCodeBlock: DockerRunCodeBlock

    lateinit var gettingStartedNavigation: GettingStartedNavigation
}

@Element(
    value = "Docker run command example",
    required = true,
    css = "div.highlight pre#__code_0"
)
class DockerRunCodeBlock : Block() {

    @Element(
        value = "Docker run code",
        required = true,
        tagName = "code"
    )
    lateinit var code: SelenideElement

    @Element(
        value = "Copy to clipboard button",
        required = true,
        css = "button.md-clipboard"
    )
    lateinit var copyButton: SelenideElement

    @Element(
        value = "Copied to clipboard message",
        css = "span.md-clipboard__message"
    )
    lateinit var copyMessage: SelenideElement
}

@Element(
    value = "Getting started navigation",
    css = "nav.md-nav--primary>ul"
)
class GettingStartedNavigation : Block() {

    @Element(
        value = "Пункты навигации",
        xpath = "./li"
    )
    lateinit var navigationItemList: Blocks<NavigationItem>

    @Step("Get navigation item with text '{textContains}'")
    fun item(textContains: String) = navigationItemList.first { it.self.has(Condition.text(textContains)) }
}

/**
 * Annotation [Element] can be placed at variable definition [GettingStartedNavigation.navigationItemList]
 */
class NavigationItem : Block() {

    @Element(
        value = "Navigation item",
        tagName = "a"
    )
    lateinit var gettingStartedNavigation: SelenideElement
}
